pkg/channel/consolidated/reconciler/controller/kafkawatcher.go

```diff
@@ -0,0 +1,185 @@
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+
+	"knative.dev/eventing-kafka/pkg/channel/consolidated/kafka"
+)
+
+var (
+	watchersMtx sync.RWMutex
+	cacheMtx    sync.RWMutex
+	// Hooks into the poll logic for testing
+	after = time.After
+	done  = func() {}
+)
+
+type CGCallBack func()
+type Matcher func(string) bool
+
+type ConsumerGroupWatcher interface {
+	// Start instructs the watcher to start polling for the consumer groups and
+	// notify any observers on the event of any changes
+	Start() error
+
+	// Watch registers callback on the event of any changes observed
+	// on the consumer groups. watcherID is an arbitrary string the user provides
+	// that will be used to identify his callbacks when provided to Forget(watcherID).
+	// Successive calls with the same watcherID will queue the callbacks successfully.
+	//
+	//To ensure this is event-triggered, level-driven,
+	// we don't pass the updates to the callback, instead the observer is expected
+	// to use List() to get the updated list of ConsumerGroups.
+	Watch(watcherID string, callback CGCallBack) error
+
+	// Forget removes all callbacks that correspond to the watcherID
+	Forget(watcherID string)
+
+	// List returns all the cached consumer groups that match matcher.
+	// It will return an empty slice if none matched or the cache is empty
+	List(matcher Matcher) []string
+}
+
+type KafkaWatcher struct {
+	logger *zap.SugaredLogger
+	//TODO name?
+	watchers     map[string][]CGCallBack
+	cgCache      []string
+	ac           kafka.AdminClient
+	pollDuration time.Duration
+}
+
+func NewKafkaWatcher(ctx context.Context, ac kafka.AdminClient, pollDuration time.Duration) KafkaWatcher {
+	w := KafkaWatcher{
+		logger:       logging.FromContext(ctx),
+		ac:           ac,
+		pollDuration: pollDuration,
+		watchers:     make(map[string][]CGCallBack),
+		cgCache:      make([]string, 0),
+	}
+	return w
+}
+
+func (w *KafkaWatcher) Start() error {
+	w.logger.Info("KafkaWatcher starting. Polling for consumer groups", zap.Duration("poll duration", w.pollDuration))
+	go func() {
+		for {
+			select {
+			case <-after(w.pollDuration):
+				cgs, err := w.ac.ListConsumerGroups()
+				if err != nil {
+					w.logger.Error("error while listing consumer groups", zap.Error(err))
+					continue
+				}
+				var notify bool
+				var cg string
+				// Look for observed CGs
+				for _, c := range cgs {
+					if !Find(w.cgCache, c) {
+						// This is the first appearance.
+						w.logger.Debug("Consumer group observed. Caching.",
+							zap.String("consumer group", c))
+						cg = c
+						notify = true
+						break
+					}
+				}
+				// Look for disappeared CGs
+				for _, c := range w.cgCache {
+					if !Find(cgs, c) {
+						// This CG was cached but it's no longer there.
+						w.logger.Debug("Consumer group deleted.",
+							zap.String("consumer group", c))
+						cg = c
+						notify = true
+						break
+					}
```

Comment:
Isn't that mostly same as https://github.com/knative-sandbox/eventing-kafka/pull/182/files#diff-85f097d6979c83d4872e73b80c62c39f59e8ef0cc922b60e251536b141b3822cR99-R106  ?

Reply:
so the first loop iterates over the new state to see which one are new (i.e are not cached), and this other loop iterates over the cached ones to see which ones no longer exist on the new state. does that make sense?

Reply:
@devguyio let's please capture this in the comments ? I am sure... this will likely cause question in the future

Reply:
@matzew sure, but I thought I did :D , did you notice the comments before each loop? or were they not clear enough? I also made the var names more descriptive. take another look and let me know if you still think it's not clear enough